### ISSUE TITLE
Get Token with Roborock App in use

### ISSUE THREAD

**Issue by edelmaca** [nicoh88/homebridge-xiaomi-roborock-vacuum#104]
Hello,
will it somehow be possible to get this plugin to work with the Roborock App in use?

If i set up the Bot (Roborock S5 Max) with the MiHome App, extract the Token and set up the Roborock App again, it won't work. I just have this one Xiaomi Roborock Device, so i have no need for the whole App Catalog of Xiaomi.

**Comment by infernalK** [comment]
You can use miHome to control your robot it works exactly the same way as the roborock one.

**Comment by edelmaca** [comment]
Yeah, I know. I admit it’s a esthetic choice. I don’t like the look of the MiHome App. It looks cheap, I don’t need the big catalog of devices, it is not completely translated and it looks like every other cheap Chinese app in the store. I don’t like the MiHome App at all. I hoped it is somehow possible to get this great plugin to work with the, in my opinion, much nicer Roborock app.
Since Roborock is building the Vacuums, I think their app will be soon the better one to use.

**Comment by afharo** [comment]
Personally I didn't even know about that app... 
Isn't there any way of extracting the token from the Roborock App? Maybe a similar process to MiHome's

**Comment by infernalK** [comment]
The 2 apps are just the same. I have tried to find the token but did succeed. And every time you change of app the token change. And it seems like you can use just one of the app at a time.

**Comment by edelmaca** [comment]
> The 2 apps are just the same. I have tried to find the token but did succeed. And every time you change of app the token change. And it seems like you can use just one of the app at a time.

You didn’t succeed with retrieving the token, Right?!
I searched for it as well, but the apps are not at all build the same, at least not on iOS. I searched for a similar file like in the MiHome file structure, but couldn’t find anything like it. But I have to say, This is not my thing. I had found the file, if it was the same ending or the same name but if I have to dig deeper, I’m lost. I would very much appreciate it if someone can find a way to get the token from the Roborock app.
The MiHome app is totally cluttered with unneeded stuff, at least for me. The Roborock App is much more stable than the MiHome App and is looking more polished. Let alone the Size difference.

**Comment by afharo** [comment]
I'll have a look, although I can't promise anything :)
I'm actually sold with your reasons @edelmaca 
I don't really like much the MiHome app. So if the Roborock App is more stable, smaller in size and more focused on the robot, I'd love to try it.

**Comment by infernalK** [comment]
I looked in the iOS backup files but didn’t find anything may be I missed it. 
I also tried to search on the net but no results.

**Comment by edelmaca** [comment]
I also searched the web. It seems no one tried to get the token from the roborock app. I found absolutely nothing related to that. Maybe the roborock app is too young and thus too less known and spread in the community.
If you need some help with testing @afharo i would happily Help you with that.

**Comment by infernalK** [comment]
Hello,
May be one of those methods can work:
https://github.com/Maxmudjon/com.xiaomi-miio/blob/master/docs/obtain_token.md

Method 4a - Packet Sender Tool
During setup of Mi Home devices the device tokens an be retrieved by sending a ping command to the device. This method uses a tool called Packet Sender which you will need to download. Choose the portable version which does not require installation.
Download the portable version of Packet Sender.
Reset the device following the instructions from the device manual, this usually means holding one or two buttons for 10 seconds. This will reset all device settings including the Wi-Fi settings.
After reset the device will create a it's own Wi-Fi network. This network will have a name related to the device and is used for configuring the device but will also allow us to retrieve the token. Connect to this Wi-Fi network with your computer which has Packet Sender running.
Open Packet Sender and enter the following details. 
HEX: 21310020ffffffffffffffffffffffffffffffffffffffffffffffffffffffff
IP: 192.168.8.1
Port: 54321
Protocol dropdown: UDP
Click send and the device will respond with an answer which contains the unique device token. In the last 16 bytes (32 characters) of the devices response is the device token. Copy and save it somewhere.
Disconnect your computer from the devices network, you can now use the Mi Home app to setup the device and connect it to your Wi-Fi network.
Method 4b - netcat and Wireshark / tcpdump
Like above you can also use this shell command to send the magic package:
echo -ne '\x21\x31\x00\x20\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff' | nc -u 192.168.8.1 54321
While running this you have to listen with Wireshark or tcpdump for UDP packages sent as anser by the robot. Extract the last 16 bytes of the answer and convert them to a (32 characters) hexadecimal string using xxd -p.
Method 5 - telnet with root access
discovered by #slavikme
In some devices, like "Mi Home Security Camera 360" (and maybe others), you are able to access the filesystem of the device using telnet.
Reset the device following the instructions from the device manual, this usually means holding one or two buttons for 10 seconds. This will reset all device settings including the Wi-Fi settings. Or you can just delete the device from Mi Home app, that will reset the device for you.
After reset the device will create it's own Wi-Fi network. This network will have a name related to the device and is used for configuring the device but will also allow us to retrieve the token. Connect to this Wi-Fi network with your computer which has a telnet command installed.
Find out Wi-Fi's gateway address (This article can help). Let's say that it is 192.168.14.1.
Create a telnet connection to this addres: 
telnet 192.168.14.1
It will ask you to enter a login. Just write root: 
Trying 192.168.14.1...
Connected to 192.168.14.1.
Escape character is '^]'.
mijia-camera login: root
#
Now you have a root shell access to device's filesystem.
Get the hex token using the following command: 
/usr/sbin/nvram get miio_token | hd -n16 -e '16/1 "%02x " "\n"' | sed '2!d; s/ //g'
The output should be something like this: 
3268786f5730305661445a6375467039
If you are not getting 32 characters hexadecimal string, then just run the command /usr/sbin/nvram get miio_token and convert it using xxd -p or this tool (don't forget to remove the space delimiter) to hexadecimal representation.
Command output example: 
# /usr/sbin/nvram get miio_token
2hxoW00VaDZcuFp9

**Comment by edelmaca** [comment]
Ok, so These methods step in before I set the bot up in one of the apps and in theory a could retrieve the token and set it up in the Roborock app after that?

I never used one of those Programmes they use to get the tokens. Looks a bit tricky to me. 🙈

**Comment by infernalK** [comment]
That’s it.
I haven’t try so I don’t know if it works.
I can try but not now may be next week.

**Comment by edelmaca** [comment]
> That’s it.
> I haven’t try so I don’t know if it works.
> I can try but not now may be next week.

Did someone try this method or has other good news?

**Comment by infernalK** [comment]
Sorry I didn’t try, not enough time. May be I’ll have some time tonight or tomorrow.

**Comment by infernalK** [comment]
Hello,
I tried this method Method 4a - Packet Sender Tool but I didn’t get the token.

**Comment by edelmaca** [comment]
Ok, thanks for your try. Now we have to hope, that someone with a bit more inside knowledge will get the attention of the Roborock App and will help us to get this tool to work with it.

**Comment by netRunner0** [comment]
Use NOX (Windows/Linux/MAC) and modified MiHome
https://ru.kapiba.ru/mihome/files/old/MiHome_5.6.55_vevs.apk
pass 37845128

Token shows in settings (for all devices)

**Comment by afharo** [comment]
But that’s from MiHome, not Roborock's specific app, right?

**Comment by infernalK** [comment]
@netRunner0 Does it show token before pairing? Cause what we need is to pair robot with roborock app not mihome and token change each time the robot is unpaired

**Comment by netRunner0** [comment]
@infernalK After, and only in mihome
But why is the roborock application necessary?

**Comment by infernalK** [comment]
It’s not necessary but some prefere this one because they don’t want/need all xiaomi ecosystem.

**Comment by c-o-m-m-a-n-d-e-r** [comment]
i think we should take a deeper look into the iobrocker adapter.
the latest version can get the token via login to the mi account.

**Comment by edelmaca** [comment]
Hm, but would this allow us to use the Roborock App afterwards? If we set the tool up with the data from the mi account and switch the robot to the Roborock app, the mi account will loose the connection as well as the MiHome app. 
Or what do I miss?

**Comment by normen** [comment]
Really after playing around with all of this I found the easiest way is to simply grab the token from a phone backup.

**Comment by infernalK** [comment]
Hello guys,
As I have some free time, I tried all the other method to get the token. None of them works with the S5max.

**Comment by edelmaca** [comment]
So there’s really no chance for us to use the Roborock App and HomeKit Integration at the Same time?! 😟 That’s sad.
So We have to Switch our Bots to that clunky Xiaomi app. 🙈 There’s no other possible solution?

**Comment by c-o-m-m-a-n-d-e-r** [comment]
Mhhh i'm using the MiHome App and Homekit at the same time ... no problems here...

**Comment by edelmaca** [comment]
Jup, the MiHome (Xiaomi) App is the problem here. We would like to use Roborocks own app and this tool in correlation. And that’s not possible so far. 
We have to hope, that someone is coming up with an new fresh idea to connect this tool and the Roborock App at the Same time.

**Comment by normen** [comment]
Really if you want the least amount of "fat" and the most flexible configuration you should go with the custom firmware and server. Then you can display the vacuum map in a simple browser window if you want.

Otherwise, look further into getting the token that the Roborock app creates. Theres nothing really that this plugin can do about the problem you set for yourself: You want to use the robot with the token generated by the RoboRock app. So the solution is to look into ways to get that token out of the app. Many people posted hints on how to approach that. You could try the way outlined to get the token from the Xiaomi app out of a backup - obviously it will be different with the Roborock app but if you find out the steps then that'll help others like you.

**Comment by edelmaca** [comment]
I am using a S5 Max. As far as I know this device isn’t possible to be jailbroken by now. So this Plugin is the only possibility We have so far.

It’s correct, we have to find the token that was created by the Roborock app. I tried searching the Backup, but I didn’t find anything that is looking like something we could grasp on. Others tried some different methods to get a Token from the app. Without any luck.
The Problem is, I am absolutely not capable of doing deeper digging. I don’t understand and can’t write any code or understand the structure or function of such processes. And I, as a few others, would be grateful if Someone with more skills in this region could help us finding a way to use this tool with the Roborock app.
Maybe this means starting a plug-in from scratch. 🤷🏻‍♂️ Who knows.

**Comment by c-o-m-m-a-n-d-e-r** [comment]
did you tried installing io.broker with the plugin ? so there you could input your login data and then the token will be shown ...

**Comment by normen** [comment]
> Maybe this means starting a plug-in from scratch. 🤷🏻‍♂️ Who knows.

Again, it has absolutely zero to do with the plugin. You just need the token. And nobody understands these things just like that, everybody needs to do research and learn.

**Comment by edelmaca** [comment]
> > Maybe this means starting a plug-in from scratch. 🤷🏻‍♂️ Who knows.
> 
> 
> 
> Again, it has absolutely zero to do with the plugin. You just need the token. And nobody understands these things just like that, everybody needs to do research and learn.

Yeah, you’re right. 🤔 The plugin isn’t the problem. But maybe someone will come up with a way to get the token from the app or even a method without the token. 
I didn’t try it with IoBroker. I just have Homebridge right now. But it looks like I have to setup IoBroker for this use case. 🤷🏻‍♂️

**Comment by philip0112** [comment]
I am using the roborock App and can get the token via the Mirobo / Miio discover command. However, I am stuck with adding the device as it show a unknown model in the discover while I am using a roborock S6 model.

**Comment by infernalK** [comment]
Can you explain how you get the token ?

**Comment by philip0112** [comment]
> Can you explain how you get the token ?

I used the mirobo / miio command when I connect to the Hotspot of the roborock. 

mirobo discover handshake 1

**Comment by infernalK** [comment]
Thanks,
If I understand we’ll :
1 reset connection 
2 the mirobo discover... in terminal 
3 set connection with the roborock app

And I’ll get the token in the terminal 

that’s it?

**Comment by normen** [comment]
When I tried to connect that way around (first get a token, then set up the roborock app) the RoboRock app would create a new token, making the token I created manually invalid again. So I don't know if it will properly work that way.

**Comment by philip0112** [comment]
> When I tried to connect that way around (first get a token, then set up the roborock app) the RoboRock app would create a new token, making the token I created manually invalid again. So I don't know if it will properly work that way.

I am not sure will it generate a new token when connect to the Roborock App, after I added to the App, in terminal I am still able to connect by using the miio / mirobo command (so I assume still use the same token), but giving the unknown model from the command.

**Comment by edelmaca** [comment]
Did anyone try this tool? 
https://github.com/prinz3nroll3/roborock-token/releases/tag/v0.0.1

I can’t get it to find my Roborocks IP. Maybe someone can get it to work.

**Comment by bluefangs** [comment]
FYI -  I was able to get the token from the log files. This requires you to have an android device. I installed an [older version of the mihome app](https://www.apkmirror.com/apk/xiaomi-inc/mihome/mihome-5-4-49-release/). The log file had the token in plain text.

**Comment by afharo** [comment]
@bluefangs That's MiHome app, right? This issue is for the Roborock app.

**Comment by bluefangs** [comment]
> @bluefangs That's MiHome app, right? This issue is for the Roborock app.

Oh no. Apoloiges. I somehow managed to overlooked that part.

**Comment by afharo** [comment]
No worries! Thank you for your message anyway :)

**Comment by Mikellas** [comment]
So no solutions to this yet right?

**Comment by ichipek** [comment]
one thing to add about Robo App (except just the looks) is that it allows now to create Siri shortcuts, MiHome doesn't...

**Comment by goodshort** [comment]
Hey if it can help, I have written down a [simple guide](https://gist.github.com/goodshort/899be31d67be6e9884c9c0fa9b232097) to retrieve the token when using an iPhone/iPad and macOS.

**Comment by Christian-CFP** [comment]
Can someone try this file
Backup: com.roborock.smart/com.roborock.smart.plist

For me it seems there are some IDs.

**Comment by Christian-CFP** [comment]
> Can someone try this file?
> iOS Backup: com.roborock.smart/com.roborock.smart.plist
> 
> For me it seems there are some IDs!

No? 
Anyone? @infernalK maybe?

**Comment by infernalK** [comment]
I haven't look at that for a while. But I can't make some test next week

**Comment by Christian-CFP** [comment]
Any news?

**Comment by redphx** [comment]
You can't use the token in Roborock app with this plugin. The app registers your vacuum on Tuya servers, not Mi Home, so it won't work.

**Comment by Christian-CFP** [comment]
Ok, so is there a way to get this plugin to work with both type of servers?

**Comment by redphx** [comment]
Not until this plugin supports Tuya.
I don't use Homebridge, but for Home Assistant you can use this integration: https://github.com/redphx/hass-roborock

**Comment by Christian-CFP** [comment]
Oh man I would but I’m useing homebridge 😞

Any intention to work together to create a hb plugin???

**Comment by afharo** [comment]
I can see [`hass-roborock`](https://github.com/redphx/hass-roborock) uses [`pyroborock`](https://github.com/89jd/pyroborock), that uses [`typaapi`](https://github.com/codetheweb/tuyapi), which is built on JS. 

I think I can explore how to implement both connection methods. If anyone else wants to try as well, PRs are welcome 😇

**Comment by redphx** [comment]
> I can see [`hass-roborock`](https://github.com/redphx/hass-roborock) uses [`pyroborock`](https://github.com/89jd/pyroborock), that uses [`typaapi`](https://github.com/codetheweb/tuyapi), which is built on JS.
> 
> I think I can explore how to implement both connection methods. If anyone else wants to try as well, PRs are welcome 😇

You could try my version of [pyroborock](https://github.com/redphx/pyroborock) which uses tinytuya (python). It should be easier to work with.

**Comment by Kohle81** [comment]
hello everyone, is it now possible to use the roborock app in parallel?  in these instructions, the use of the roborock app is mentioned after the token has been read?

https://www.simon42.com/roborock-homekit-token-einfach/

**Comment by torandreroland** [comment]
Hi! I have the Roborock app installed on my MacBook Air M1. I was sneaking around in the container data, and more specifically ~/Library/Containers/Roborock/Data/Library/Preferences/com.roborock.smart.plist. Anybody making sense of this?  A bit hesitant to share the information as I suspect it includes the necessary tokens etc...

**Comment by sohsatoh** [comment]
From looking at the Roborock app on iOS it looks like it is using React Native to communicate with server (maybe local too).
There is a file called "index.ios.bundle" in the app's documentation folder, but since the app is verifying the signature in "index.ios.bundle.sign", it is not possible to rewrite this file. (If you are jailbroken, it's easy to bypass it).
Also, in RRPluginSDK class, there is a selector -(void)getVoicePackageList: (id)arg2 resolver: (id)arg3 rejecter: (id)arg4, which I tried to hook, but it was never called, so it seems that all communication is done through React Native.
Unfortunately I don't have any knowledge of React Native so I'm not going to test this further, but I'm happy to help anyone out.

**Comment by chris400** [comment]
@goodshort your guide seems to be only for Xiaomi Home App, not the "Roborock App" right? The whole topic is about the latter :\

**Comment by chris400** [comment]
@edelmaca pls consider editing the issue title. The comments - how to extract token with Xiaomi App - keep appearing on and on :)

**Comment by Christian-CFP** [comment]
@chris400 The plugin works just with the Xiaomi app. Both of them uses different servers and api's. So until now it just works with Xiaomi.

Even I would be greatfull if it works with the Roborock App it self because I just need the cleaning app not the entire Xiaomi App.

**Comment by goodshort** [comment]
> @goodshort your guide seems to be only for Xiaomi Home App, not the "Roborock App" right? The whole topic is about the latter :\
> 
> 

Just edited my comment. Thanks for the heads up!

**Comment by leroi21421** [comment]
Hey! 
any update about it? 
did you manage to get the token for the roborock app? If so - please write down the way to do it🙏🏻

**Comment by Newbie2Quick** [comment]
Okay, so I am new here. But been digging around for the same question... "how to find the token in the RoboRock App?" 
..but after reading this thread, I saw someone bring up even if we find the token from the roborock app, the plugin for HomeBridge only works with Xiaomi servers and not Tuya servers.... 1) is this true? 2) If I found the token, is there away around this? 

So the good news. I found the token in the roborock app / backup files ripped from an iOS device (iphone). Using a program called iBackup Viewer. 
Its located in the following file: 
_Library/Preferences/com.roborock.smart.plist_  

Then located under:
_UserDefaultKeyUserModel_

Sub:
_token_

With this information, Im thinking two things can happen. 

1. figure out a work around
2. write a new plugin

All input is welcomed! :)

**Comment by Einstein2150** [comment]
Have you tried to control the robot over the miio-api with the extracted token? This is the first thing I would try.

**Comment by Newbie2Quick** [comment]
@Einstein2150 ... what is the miio-api? ... referring to the Xiaomi Homebridge plugin?  If so, I tried. no luck... but wondering if I'm typing the right token... let me explain...

The token l found has 32 characters, then a dash (-) followed by 40 characters. I noticed from looking around the 40 character is the same as the OpenUDID. 

So, do use all 72 characters, with or without the dash for the token? OR do I just use 32C. or 40C. ?  

Also, I found a YouTube video of a guy, ripping a token from a backup file, but had to decrypt it using a plug and play web based tool. ... is it possible this needs to be decrypted?

**Comment by Newbie2Quick** [comment]
Just curios... wonder if there is a work around using either IFTTT, Shortcuts, or both... 

Example: I have a wall mounted Ipad. lets say its logged into the RoboRock App. wonder if you could create an relay, from Homekit>homebridge>IFTTT>Roborock  OR something similar...

**Comment by Einstein2150** [comment]
> @Einstein2150 ... what is the miio-api? ... referring to the Xiaomi Homebridge plugin?  If so, I tried. no luck... but wondering if I'm typing the right token... let me explain...
> 
> 
> 
> The token l found has 32 characters, then a dash (-) followed by 40 characters. I noticed from looking around the 40 character is the same as the OpenUDID. 
> 
> 
> 
> So, do use all 72 characters, with or without the dash for the token? OR do I just use 32C. or 40C. ?  
> 
> 
> 
> Also, I found a YouTube video of a guy, ripping a token from a backup file, but had to decrypt it using a plug and play web based tool. ... is it possible this needs to be decrypted? 
> 
> 
> 
> 

The token is normally AES encrypted with the key of 32 zeros.

Miio is a GitHub project to communicate with Mi devices. Here is the link to the project [https://github.com/rytilahti/python-miio]() You can install it with pip: `pip3 install python-miio`

**Comment by Newbie2Quick** [comment]
... unfortunately I do not know how to install it using pip. 

do you know of a link or a quick walk through?

**Comment by Einstein2150** [comment]
> ... unfortunately I do not know how to install it using pip. 
> 
> 
> 
> do you know of a link or a quick walk through?

pip is a part of python3. Miio is written in Python

**Comment by Newbie2Quick** [comment]
Okay. I am obviously not an expert lol 

I guess I am asking, how to install https://github.com/rytilahti/python-miio.git onto my home bridge. I tried using the terminal and calling for an install from the github link. but that did not work ... received username and password error. I tried adding "sudo" and still no luck. I feel like installing a github code onto homebridge should be easy, but I must be missing something.

**Comment by Newbie2Quick** [comment]
I just tried "pip install https://github.com/rytilahti/python-miio.git" and it started to download, then errored out

**Comment by Newbie2Quick** [comment]
add git+.... seems to be installing

**Comment by Newbie2Quick** [comment]
So i have successfully installed python-miio ... but I do not know how to add the vacuum :(

**Comment by Newbie2Quick** [comment]
I tried this...

vac = Vacuum("<ip address>", "<token>")

but getting syntax error near unexpected token `('

**Comment by Newbie2Quick** [comment]
@Einstein2150 

just found this:

_> Roborock Vacuum not detected
It seems that a Roborock vacuum connected through the Roborock app (and not the Xiaomi Home app) won’t allow control over local network,
even with a valid token, leading to the following exception:

mirobo.device.DeviceException: Unable to discover the device x.x.x.x
Resetting the device’s wifi and pairing it again with the Xiaomi Home app should solve the issue._

**Comment by heiko7even** [comment]
Hey there :-) Is there anything new on this?

**Comment by JeromeAstero** [comment]
Good day!

I got localkey and device id using Charles(ios).
Instruction here:
[Easy way to obtain tokens using iOS](https://github.com/Maxmudjon/com.xiaomi-miio/issues/133)

Localkey is not accepted as a token by the plugin.

Does anyone have any idea how to convert localkey to token?

**Comment by JustAnotherPU** [comment]
any news about the token and tuya support for the homebridge plugin ?

**Comment by alesp56** [comment]
Is anything new about the token retrieval with the Roborock app?

**Comment by Frederick-G764** [comment]
Can you tell us @4c0d3r how you did this?

**Comment by 4c0d3r** [comment]
@Frederick-G764 follow the link, run this script: https://gist.github.com/rovo89/dff47ed19fca0dfdda77503e66c2b7c7 with node.

```node login.js username password```

Didn't write the script, so not an expert on it.

**Comment by FezVrasta** [comment]
@4c0d3r do you know which token should I use after I run that script? I tried with `userdata.json` -> `token` but the plugin fails to connect to the robot.

I also tried with `homedata.json` -> `localKey` with the same issue.

**Comment by 4c0d3r** [comment]
I'm pretty sure the plugin urls need to change.   One set of apis is run by roborock and the other is run by xaiomi.

**Comment by JeromeAstero** [comment]
> Roborock user tokens can be easily had now with: [rytilahti/python-miio#1572](https://github.com/rytilahti/python-miio/discussions/1572)

Can you please tell me how this tool solves the problem of extracting a token for use in HomeBridge?
I ran the script, got userdata.json, homedata.json but none of the keys are suitable for use as a token.

**Comment by FezVrasta** [comment]
Is anyone writing a homebridge port of https://github.com/humbertogontijo/homeassistant-roborock? In case please get in touch with me, I'd like to contribute.

**Comment by o-mega** [comment]
Did anyone succeed to get token with node script above?
I run the script, got two files with my roborock data (even details about the vacuum), but I cannot find token that fit homebridge plugin

**Comment by FezVrasta** [comment]
Tokens are different.

**Comment by o-mega** [comment]
So, is there a way to connect roborock app, roborock s7 and homebridge together ? I really don't want to use mi home instead of roborock app 😞

**Comment by FezVrasta** [comment]
Only through Home Assistant for now, please, read the comments before yours.

**Comment by mdxl** [comment]
> Only through Home Assistant for now, please, read the comments before yours.

I installed this integration - https://github.com/humbertogontijo/homeassistant-roborock, but I can't find the token anywhere. In which menu can I find this information?

**Comment by dersparkiie** [comment]
> > Only through Home Assistant for now, please, read the comments before yours.
> 
> I installed this integration - https://github.com/humbertogontijo/homeassistant-roborock, but I can't find the token anywhere. In which menu can I find this information?

You have to activate debug logging for the integration. Than you'll find the token when Restarting Home Assistant in the log ;)

**Comment by clemone210** [comment]
> The code from rovo's gist works. I got it about 1 year ago from https://forum.iobroker.net/topic/56098/tester-gesucht-roborock-api
> 
> Tweaked it and has been working for that time using Homekit automation to trigger when to clean based on if people are at home or leaving.
> 
> Cant provide code as it is already a spaghetti code with my secrets.
> 
> I used https://github.com/staromeste/homebridge-http-advanced-accessory#readme to call my nodejs server powered by the script from rovo.
> 
> Once you start investigating the responses from the mqtt of roborock and find some that are missing in the official docs you can make an unencrypted backup of your iOS device. Then you can inspect the file system of the App for the minified Javascript files that make the Roborock app (because its react based). Here is an example of code used in their App that you can use to map "state code" to "state description":
> 
> ```js
> A = {
>         UNKNOWN: "UNKNOWN",
>         INITIAL: "INITIAL",
>         SLEEPING: "SLEEPING",
>         WAITING: "WAITING",
>         REMOTE: "REMOTE",
>         CLEAN: "CLEAN",
>         BACK_TO_DOCK: "BACK_TO_DOCK",
>         CHARGING: "CHARGING",
>         CHARGE_ERROR: "CHARGE_ERROR",
>         FULL_CHARGE: "FULL_CHARGE",
>         PAUSE: "PAUSE",
>         SPOT_CLEAN: "SPOT_CLEAN",
>         MALFUNCTIONING: "MALFUNCTIONING",
>         PREPARE_SHUTDOWN: "PREPARE_SHUTDOWN",
>         UPDATING: "UPDATING",
>         RUB_TO_DOCK: "RUB_TO_DOCK",
>         GOTO_TARGET: "GOTO_TARGET",
>         ZONED_CLEAN: "ZONED_CLEAN",
>         SEGMENT_CLEAN: "SEGMENT_CLEAN",
>         OFF_LINE: "OFF_LINE",
>         LOCKED: "LOCKED",
>         COLLECTING_DUST: "COLLECTING_DUST",
>         WASHING_DUSTER: "WASHING_DUSTER",
>         MOPPING: "MOPPING",
>         CLEAN_MOP_CLEANING: "CLEAN_MOP_CLEANING",
>         CLEAN_MOP_MOPPING: "CLEAN_MOP_MOPPING",
>         SEGMENT_MOPPING: "SEGMENT_MOPPING",
>         SEGMENT_CLEAN_MOP_CLEANING: "SEGMENT_CLEAN_MOP_CLEANING",
>         SEGMENT_CLEAN_MOP_MOPPING: "SEGMENT_CLEAN_MOP_MOPPING",
>         ZONED_MOPPING: "ZONED_MOPPING",
>         ZONED_CLEAN_MOP_CLEANING: "ZONED_CLEAN_MOP_CLEANING",
>         ZONED_CLEAN_MOP_MOPPING: "ZONED_CLEAN_MOP_MOPPING",
>         BACK_TO_DOCK_WASHING_DUSTER: "BACK_TO_DOCK_WASHING_DUSTER",
>         AIR_DRYING_STOPPING: "AIR_DRYING_STOPPING",
>         VOICE_CHATTING: "VOICE_CHATTING",
>         QUICK_BUILDING_MAP: "QUICK_BUILDING_MAP",
>         EGG_ATTACK: "EGG_ATTACK",
>       };
>     e.RobotState = A;
>     (e.RobotCompactState = {
>       UNKNOWN: "UNKNOWN",
>       INITIAL: "INITIAL",
>       SLEEPING: "SLEEPING",
>       WAITING: "WAITING",
>       REMOTE: "REMOTE",
>       CLEAN: "CLEAN",
>       BACK_TO_DOCK: "BACK_TO_DOCK",
>       CHARGING: "CHARGING",
>       CHARGE_ERROR: "CHARGE_ERROR",
>       FULL_CHARGE: "FULL_CHARGE",
>       PAUSE: "PAUSE",
>       SPOT_CLEAN: "SPOT_CLEAN",
>       MALFUNCTIONING: "MALFUNCTIONING",
>       PREPARE_SHUTDOWN: "PREPARE_SHUTDOWN",
>       UPDATING: "UPDATING",
>       RUB_TO_DOCK: "RUB_TO_DOCK",
>       GOTO_TARGET: "GOTO_TARGET",
>       ZONED_CLEAN: "ZONED_CLEAN",
>       SEGMENT_CLEAN: "SEGMENT_CLEAN",
>       OFF_LINE: "OFF_LINE",
>     }),
>       Object.freeze(A);
>     var G = {
>       0: { name: v.localization_strings_Common_Constants_0, value: A.UNKNOWN },
>       1: { name: v.localization_strings_Common_Constants_1, value: A.INITIAL },
>       2: { name: v.localization_strings_Common_Constants_2, value: A.SLEEPING },
>       3: { name: v.localization_strings_Common_Constants_3, value: A.WAITING },
>       7: { name: v.localization_strings_Common_Constants_4, value: A.REMOTE },
>       5: { name: v.localization_strings_Common_Constants_1, value: A.CLEAN },
>       6: {
>         name: v.localization_strings_Common_Constants_6,
>         value: A.BACK_TO_DOCK,
>       },
>       8: { name: v.localization_strings_Common_Constants_8, value: A.CHARGING },
>       9: {
>         name: v.localization_strings_Common_Constants_10,
>         value: A.CHARGE_ERROR,
>       },
>       10: { name: v.localization_strings_Common_Constants_11, value: A.PAUSE },
>       11: {
>         name: v.localization_strings_Common_Constants_12,
>         value: A.SPOT_CLEAN,
>       },
>       12: {
>         name: v.localization_strings_Common_Constants_13,
>         value: A.MALFUNCTIONING,
>       },
>       14: {
>         name: v.localization_strings_Common_Constants_15,
>         value: A.UPDATING,
>       },
>       15: {
>         name: v.localization_strings_Common_Constants_6,
>         value: A.RUB_TO_DOCK,
>       },
>       16: { name: v.rubys_main_subtitle_goto_target, value: A.GOTO_TARGET },
>       17: { name: v.rubys_main_button_text_zone, value: A.ZONED_CLEAN },
>       18: { name: v.rubys_main_segment_clean_alert, value: A.SEGMENT_CLEAN },
>       22: { name: v.robot_status_collecting_dust, value: A.COLLECTING_DUST },
>       23: { name: v.robot_status_washing_duster, value: A.WASHING_DUSTER },
>       25: { name: v.robot_status_washing_duster, value: A.WASHING_DUSTER },
>       26: {
>         name: v.robot_status_back_dock_washing_duster,
>         value: A.BACK_TO_DOCK_WASHING_DUSTER,
>       },
>       28: { name: v.be_on_the_phone, value: A.VOICE_CHATTING },
>       29: {
>         name: v.robot_status_quick_building_map,
>         value: A.QUICK_BUILDING_MAP,
>       },
>       30: { name: v.robot_status_egg_attack, value: A.EGG_ATTACK },
>       100: {
>         name: v.localization_strings_Common_Constants_17,
>         value: A.FULL_CHARGE,
>       },
>       101: {
>         name: v.localization_strings_Common_Constants_18,
>         value: A.OFF_LINE,
>       },
>       102: {
>         name: v.localization_strings_Common_Constants_20,
>         value: A.UNKNOW,
>       },
>       103: { name: v.robot_status_locked_when_saving_map, value: A.LOCKED },
>       202: {
>         name: v.robot_status_air_drying_stopping,
>         value: A.AIR_DRYING_STOPPING,
>       },
>       6301: { name: v.robot_status_mopping, value: A.MOPPING },
>       6302: {
>         name: v.localization_strings_Common_Constants_1,
>         value: A.CLEAN_MOP_CLEANING,
>       },
>       6303: { name: v.robot_status_mopping, value: A.CLEAN_MOP_MOPPING },
>       6304: { name: v.robot_status_segment_mopping, value: A.SEGMENT_MOPPING },
>       6305: {
>         name: v.rubys_main_segment_clean_alert,
>         value: A.SEGMENT_CLEAN_MOP_CLEANING,
>       },
>       6306: {
>         name: v.robot_status_segment_mopping,
>         value: A.SEGMENT_CLEAN_MOP_MOPPING,
>       },
>       6307: { name: v.robot_status_zone_mopping, value: A.ZONED_MOPPING },
>       6308: {
>         name: v.rubys_main_button_text_zone,
>         value: A.ZONED_CLEAN_MOP_CLEANING,
>       },
>       6309: {
>         name: v.robot_status_zone_mopping,
>         value: A.ZONED_CLEAN_MOP_MOPPING,
>       },
>       6310: {
>         name: v.robot_status_back_dock_washing_duster,
>         value: A.BACK_TO_DOCK_WASHING_DUSTER,
>       },
>     };
> ```

would you be able to remove your secrets and share how to do it?

**Comment by jstraw4663** [comment]
Any luck here? I'm curious as to how they can do it through Home Assistant, does anyone have contacts with the dev for the HA project? 

Sounds like it should work as long as we get the Roborock token right?

**Comment by rhusar** [comment]
Before I unsubscribe from this topic, the PSA is that HomeAssistant has released 2023.5 with the Roborock plugin https://www.home-assistant.io/integrations/roborock/ which works flawlessly with Roborock credentials (without any Xiaomi / MiHome crap). So official Roborock app + switch in Apple Home via HomeAssistant. Done!

**Comment by austinarchibald** [comment]
Following this. Want to get token from Roborock app (not Xiaomi Cloud) and use in Homebridge. Doesn't look like there's a solution yet. Home Assistant only.

**Comment by hroland** [comment]
anyone yet to take a stab at looking at home-assistant's Python code? :)

https://github.com/home-assistant/core/tree/dev/homeassistant/components/roborock

maybe transpile it over to JS with some help from GPT? :)

**Comment by whatUwant** [comment]
Could this be helpful? https://gist.github.com/rovo89/dff47ed19fca0dfdda77503e66c2b7c7

**Comment by tasict** [comment]
https://github.com/copystring/ioBroker.roborock

@afharo 
I found an iobroker plug-in for Roborock vacuums, and it's written in Nodejs.
These resources might be helpful for you.

MiHome has been slow to update its support for new vacuums, and many new models seem to be unsupported. So your help is needed.

**Comment by afharo** [comment]
@tasict, thanks! I'll take a look as soon as my kids spare me some "me" time 😅

If anyone around wants to give it a try before I get to jump into it, I'll be happy to review their PR :)

**Comment by tasict** [comment]
> @tasict, thanks! I'll take a look as soon as my kids spare me some "me" time 😅
> 
> If anyone around wants to give it a try before I get to jump into it, I'll be happy to review their PR :)

Haha~
Actually, I also tried to clone your code and add support for the Roborock App from it, but like you, I have to find time from my limited free time outside of taking care of my children to try and make modifications. Otherwise, my wife might be unhappy, so the progress is quite slow.

**Comment by spilger** [comment]
Did anyone of you look into the authentication? Looks like they are using hawk authentication in the Roborock app, which means, there would need to be a new way to authenticate ein the plugin

**Comment by blazejpawlak** [comment]
It seems that the Roborock app token is different from the Mi Home app as outlined [here](https://github.com/redphx/roborock-tuya-token/issues/2#issuecomment-806223667).

So even if you're able to extract the token (see [this](https://github.com/redphx/roborock-tuya-token?tab=readme-ov-file), and [this](https://community.home-assistant.io/t/custom-component-roborock-communication/229032/155) you won't be able to use it with this plugin. It seem that this plugin would need to support the [Tuya protocol](https://codetheweb.github.io/tuyapi/) first.

Does anyone know what are the major differences in features of both apps?

**Comment by tasict** [comment]
@blazejpawlak The main reason is that compared to the Roborock app, the Mi Home app has stopped updating for new models of Roborock vacuum cleaners. Therefore, models released last year cannot be used with Mi Home.

**Comment by blazejpawlak** [comment]
I was able to retrieve the token Roborock token using this [script](https://gist.github.com/rovo89/dff47ed19fca0dfdda77503e66c2b7c7). Haven't had the chance to test it yet though.
There is also [this](https://python-roborock.readthedocs.io/en/latest/usage.html) python based project that works pretty well.

**Comment by afharo** [comment]
There's also this [Home Assistant integration](https://www.home-assistant.io/integrations/roborock/). Maybe some of you prefer to try that other approach instead of homebridge.

**Comment by gino8080** [comment]
> I was able to retrieve the token Roborock token using this...  [this](https://python-roborock.readthedocs.io/en/latest/usage.html) python based project that works pretty well.

this one worked for me thankyou!

**Comment by dandek01** [comment]
> > I was able to retrieve the token Roborock token using this...  [this](https://python-roborock.readthedocs.io/en/latest/usage.html) python based project that works pretty well.
> 
> this one worked for me thankyou!

Hi, how did you manage to extract token using python-roborock ?

**Comment by thibault-carpentier** [comment]
> > I was able to retrieve the token Roborock token using this...  [this](https://python-roborock.readthedocs.io/en/latest/usage.html) python based project that works pretty well.
> 
> this one worked for me thankyou!

@gino8080: I was able to get the token from the script but no luck using it in homebridge with this plugin. Am I missing something?

**Comment by tasict** [comment]
I think it's because of the difference in login regions, the default code connects to the European cloud server URL. You need to try modifying the code here. Perhaps you can change it to the US cloud server (https://usiot.roborock.com/) and see if that works: 
https://github.com/humbertogontijo/python-roborock/blob/2731bcea38bd3bf73f84a11f17df6be59ae2637f/roborock/web_api.py#L35

In fact, I’ve already started researching this library and have begun implementing a version for Homebridge. Basic control has already been achieved. But since it's just the beginning, the code is still quite messy, and I’m working on organizing it. Originally, I was planning to wait until the HomeKit vacuum category (iOS 18.1??) is available, then I will release it for everyone to reference or integrate into the original code of homebridge-xiaomi-roborock-vacuum.

**Comment by thibault-carpentier** [comment]
@tasict Thanks for the answer. I tried changing the server, but I got back the same token. It's a 22-char long token and I believe the miio tokens are all 32-char long, so I don't believe they would work out of the box.

Looking forward to the updates!

**Comment by tasict** [comment]
As previously discussed, I referred to the communication protocol from https://github.com/copystring/ioBroker.roborock. Initially, I intended to modify homebridge-xiaomi-roborock-vacuum to support the communication protocol used directly by the Roborock app. However, since I was not very familiar with the architecture of homebridge-xiaomi-roborock-vacuum, I was concerned about disrupting the existing user experience.

Therefore, I attempted to develop [homebridge-roborock-vacuum](https://github.com/tasict/homebridge-roborock-vacuum), a Homebridge plugin that directly uses the Roborock app's communication protocol. Since this is just a prototype, I deliberately simplified many features, leaving only basic on/off control and status reporting.

For those interested in testing it, you can search for it directly in the Homebridge interface. I have tested it with the two robot vacuum models I own: Roborock Qrevo MaxV and Roborock Qrevo. Notably, the original communication protocol did not support the Roborock Qrevo MaxV, but it now works with my plugin.

I hope that someone can help integrate my implementation into homebridge-xiaomi-roborock-vacuum in the future.
